File: C++/141_Linked List Cycle.cpp
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == NULL || head->next == NULL)
            return false;
        ListNode *slow = head, *fast = head->next;
        while (fast != NULL && slow != fast) {
            fast = fast->next;
            fast = fast == NULL ? fast : fast->next;
            slow = slow->next;
        }
        return slow == fast;
    }
};

File: Python/141_Linked List Cycle.py
class Solution:
    # @param head, a ListNode
    # @return a boolean
    def hasCycle(self, head):
        if not head:
            return False
        if head.next == head:
            return True
        second = head.next
        first = head
        while second and second != first:
            second = second.next
            if second:
                second = second.next
            first = first.next
        return second == first